File: go/json_output_test.go
package ccc

import (
	"testing"
)

func TestParseOpenCodeJson_Response(t *testing.T) {
	raw := `{"response":"hello world"}` + "\n"
	out := ParseOpenCodeJson(raw)
	if out.SchemaName != "opencode" {
		t.Fatalf("schema: %q", out.SchemaName)
	}
	if out.FinalText != "hello world" {
		t.Fatalf("final: %q", out.FinalText)
	}
	if len(out.Events) != 1 || out.Events[0].EventType != "text" {
		t.Fatalf("events: %+v", out.Events)
	}
}

func TestParseOpenCodeJson_Error(t *testing.T) {
	raw := `{"error":"boom"}` + "\n"
	out := ParseOpenCodeJson(raw)
	if out.Error != "boom" {
		t.Fatalf("error: %q", out.Error)
	}
	if len(out.Events) != 1 || out.Events[0].EventType != "error" {
		t.Fatalf("events: %+v", out.Events)
	}
}

func TestParseOpenCodeJson_MixedLines(t *testing.T) {
	raw := `{"response":"hi"}` + "\n" + `{"response":"bye"}` + "\n"
	out := ParseOpenCodeJson(raw)
	if out.FinalText != "bye" {
		t.Fatalf("final: %q", out.FinalText)
	}
	if len(out.Events) != 2 {
		t.Fatalf("events: %d", len(out.Events))
	}
}

func TestParseOpenCodeJson_SkipsInvalidJSON(t *testing.T) {
	raw := "not json\n" + `{"response":"ok"}` + "\n"
	out := ParseOpenCodeJson(raw)
	if out.FinalText != "ok" {
		t.Fatalf("final: %q", out.FinalText)
	}
	if len(out.RawLines) != 1 {
		t.Fatalf("raw_lines: %d", len(out.RawLines))
	}
}

func TestParseClaudeCodeJson_Init(t *testing.T) {
	raw := `{"type":"system","subtype":"init","session_id":"sess123"}` + "\n"
	out := ParseClaudeCodeJson(raw)
	if out.SessionID != "sess123" {
		t.Fatalf("session: %q", out.SessionID)
	}
}

func TestParseClaudeCodeJson_Assistant(t *testing.T) {
	raw := `{"type":"assistant","message":{"content":[{"type":"text","text":"hello"}],"usage":{"input_tokens":10,"output_tokens":5}}}` + "\n"
	out := ParseClaudeCodeJson(raw)
	if out.FinalText != "hello" {
		t.Fatalf("final: %q", out.FinalText)
	}
	if out.Usage["input_tokens"] != 10 {
		t.Fatalf("usage: %+v", out.Usage)
	}
	if len(out.Events) != 1 || out.Events[0].EventType != "assistant" {
		t.Fatalf("events: %+v", out.Events)
	}
}

func TestParseClaudeCodeJson_ToolUse(t *testing.T) {
	raw := `{"type":"tool_use","tool_name":"read","tool_input":{"path":"/tmp/x"}}` + "\n"
	out := ParseClaudeCodeJson(raw)
	if len(out.Events) != 1 {
		t.Fatalf("events: %d", len(out.Events))
	}
	ev := out.Events[0]
	if ev.EventType != "tool_use" || ev.ToolCall == nil {
		t.Fatalf("bad event: %+v", ev)
	}
	if ev.ToolCall.Name != "read" {
		t.Fatalf("tool name: %q", ev.ToolCall.Name)
	}
	if ev.ToolCall.Arguments != `{"path":"/tmp/x"}` {
		t.Fatalf("tool args: %q", ev.ToolCall.Arguments)
	}
}

func TestParseClaudeCodeJson_ToolResult(t *testing.T) {
	raw := `{"type":"tool_result","tool_use_id":"tc1","content":"file contents","is_error":false}` + "\n"
	out := ParseClaudeCodeJson(raw)
	ev := out.Events[0]
	if ev.EventType != "tool_result" || ev.ToolResult == nil {
		t.Fatalf("bad event: %+v", ev)
	}
	if ev.ToolResult.ToolCallID != "tc1" || ev.ToolResult.Content != "file contents" {
		t.Fatalf("tool_result: %+v", ev.ToolResult)
	}
}

func TestParseClaudeCodeJson_ResultSuccess(t *testing.T) {
	raw := `{"type":"result","subtype":"success","result":"done","cost_usd":0.05,"duration_ms":1200}` + "\n"
	out := ParseClaudeCodeJson(raw)
	if out.FinalText != "done" {
		t.Fatalf("final: %q", out.FinalText)
	}
	if out.CostUSD != 0.05 {
		t.Fatalf("cost: %f", out.CostUSD)
	}
	if out.DurationMs != 1200 {
		t.Fatalf("duration: %d", out.DurationMs)
	}
}

func TestParseClaudeCodeJson_ResultError(t *testing.T) {
	raw := `{"type":"result","subtype":"error","error":"fail"}` + "\n"
	out := ParseClaudeCodeJson(raw)
	if out.Error != "fail" {
		t.Fatalf("error: %q", out.Error)
	}
	if len(out.Events) != 1 || out.Events[0].EventType != "error" {
		t.Fatalf("events: %+v", out.Events)
	}
}

func TestParseClaudeCodeJson_ThinkingDelta(t *testing.T) {
	raw := `{"type":"stream_event","event":{"type":"content_block_delta","delta":{"type":"thinking_delta","thinking":"hmm"}}}` + "\n"
	out := ParseClaudeCodeJson(raw)
	ev := out.Events[0]
	if ev.EventType != "thinking_delta" || ev.Thinking != "hmm" {
		t.Fatalf("event: %+v", ev)
	}
}

func TestParseKimiJson_AssistantString(t *testing.T) {
	raw := `{"role":"assistant","content":"hi there"}` + "\n"
	out := ParseKimiJson(raw)
	if out.FinalText != "hi there" {
		t.Fatalf("final: %q", out.FinalText)
	}
}

func TestParseKimiJson_AssistantContentList(t *testing.T) {
	raw := `{"role":"assistant","content":[{"type":"text","text":"hello"},{"type":"think","think":"pondering"},{"type":"text","text":"world"}]}` + "\n"
	out := ParseKimiJson(raw)
	if out.FinalText != "hello\nworld" {
		t.Fatalf("final: %q", out.FinalText)
	}
	var thinkFound bool
	for _, ev := range out.Events {
		if ev.EventType == "thinking" {
			thinkFound = true
			if ev.Thinking != "pondering" {
				t.Fatalf("thinking: %q", ev.Thinking)
			}
		}
	}
	if !thinkFound {
		t.Fatal("no thinking event found")
	}
}

func TestParseKimiJson_ToolCall(t *testing.T) {
	raw := `{"role":"assistant","content":"","tool_calls":[{"id":"tc1","function":{"name":"bash","arguments":"{\"cmd\":\"ls\"}"}}]}` + "\n"
	out := ParseKimiJson(raw)
	if len(out.Events) != 2 {
		t.Fatalf("events: %d", len(out.Events))
	}
	ev := out.Events[1]
	if ev.EventType != "tool_call" || ev.ToolCall == nil {
		t.Fatalf("event: %+v", ev)
	}
	if ev.ToolCall.Name != "bash" || ev.ToolCall.Arguments != `{"cmd":"ls"}` {
		t.Fatalf("tool_call: %+v", ev.ToolCall)
	}
}

func TestParseKimiJson_ToolResult(t *testing.T) {
	raw := `{"role":"tool","tool_call_id":"tc1","content":[{"type":"text","text":"output line"},{"type":"text","text":"<system>hidden</system>"}]}` + "\n"
	out := ParseKimiJson(raw)
	ev := out.Events[0]
	if ev.EventType != "tool_result" || ev.ToolResult == nil {
		t.Fatalf("event: %+v", ev)
	}
	if ev.ToolResult.Content != "output line" {
		t.Fatalf("content: %q", ev.ToolResult.Content)
	}
	if ev.ToolResult.ToolCallID != "tc1" {
		t.Fatalf("tool_call_id: %q", ev.ToolResult.ToolCallID)
	}
}

func TestParseKimiJson_PassthroughEvents(t *testing.T) {
	raw := `{"type":"TurnBegin"}` + "\n" + `{"type":"StepBegin"}` + "\n" + `{"type":"StatusUpdate"}` + "\n"
	out := ParseKimiJson(raw)
	if len(out.Events) != 3 {
		t.Fatalf("events: %d", len(out.Events))
	}
	if out.Events[0].EventType != "turnbegin" {
		t.Fatalf("type[0]: %q", out.Events[0].EventType)
	}
	if out.Events[1].EventType != "stepbegin" {
		t.Fatalf("type[1]: %q", out.Events[1].EventType)
	}
	if out.Events[2].EventType != "statusupdate" {
		t.Fatalf("type[2]: %q", out.Events[2].EventType)
	}
}

func TestParseJsonOutput_Dispatch(t *testing.T) {
	out := ParseJsonOutput(`{"response":"hi"}`+"\n", "opencode")
	if out.SchemaName != "opencode" || out.FinalText != "hi" {
		t.Fatalf("bad dispatch: %+v", out)
	}
}

func TestParseJsonOutput_UnknownSchema(t *testing.T) {
	out := ParseJsonOutput("", "unknown")
	if out.Error == "" {
		t.Fatal("expected error for unknown schema")
	}
}

func TestRenderParsed_Text(t *testing.T) {
	out := ParsedJsonOutput{
		FinalText: "final",
		Events: []JsonEvent{
			{EventType: "assistant", Text: "hello"},
			{EventType: "assistant", Text: "world"},
		},
	}
	s := RenderParsed(out)
	if s != "hello\nworld" {
		t.Fatalf("rendered: %q", s)
	}
}

func TestRenderParsed_ToolUseAndResult(t *testing.T) {
	out := ParsedJsonOutput{
		Events: []JsonEvent{
			{EventType: "tool_use", ToolCall: &ToolCall{Name: "bash"}},
			{EventType: "tool_result", ToolResult: &ToolResult{Content: "ok"}},
		},
	}
	s := RenderParsed(out)
	if s != "[tool] bash\n[tool_result] ok" {
		t.Fatalf("rendered: %q", s)
	}
}

func TestRenderParsed_Thinking(t *testing.T) {
	out := ParsedJsonOutput{
		Events: []JsonEvent{
			{EventType: "thinking_delta", Thinking: "deep thoughts"},
		},
	}
	s := RenderParsed(out)
	if s != "[thinking] deep thoughts" {
		t.Fatalf("rendered: %q", s)
	}
}

func TestRenderParsed_EmptyFallsBackToFinalText(t *testing.T) {
	out := ParsedJsonOutput{FinalText: "fallback"}
	s := RenderParsed(out)
	if s != "fallback" {
		t.Fatalf("rendered: %q", s)
	}
}

func TestRenderParsed_Error(t *testing.T) {
	out := ParsedJsonOutput{
		Events: []JsonEvent{
			{EventType: "error", Text: "something broke"},
		},
	}
	s := RenderParsed(out)
	if s != "[error] something broke" {
		t.Fatalf("rendered: %q", s)
	}
}

File: php/tests/JsonOutputTest.php
#!/usr/bin/env php
<?php

require_once __DIR__ . '/../src/JsonOutput.php';

use Call\Coding\Clis\JsonOutput;
use Call\Coding\Clis\ParsedJsonOutput;

$passed = 0;
$failed = 0;

function assert_test(string $name, bool $condition): void
{
    global $passed, $failed;
    if ($condition) {
        $passed++;
        echo "ok - $name\n";
    } else {
        $failed++;
        echo "not ok - $name\n";
    }
}

$r = JsonOutput::parseOpencodeJson("{\"response\": \"hello\"}\n");
assert_test('opencode response event', $r->events[0]->eventType === 'text' && $r->events[0]->text === 'hello');
assert_test('opencode final text', $r->finalText === 'hello');

$r = JsonOutput::parseOpencodeJson("{\"error\": \"fail\"}\n");
assert_test('opencode error', $r->error === 'fail' && $r->events[0]->eventType === 'error');

$r = JsonOutput::parseOpencodeJson("bad\n{\"response\": \"ok\"}\n");
assert_test('opencode skips invalid', $r->finalText === 'ok' && count($r->events) === 1);

$r = JsonOutput::parseClaudeCodeJson("{\"type\":\"system\",\"subtype\":\"init\",\"session_id\":\"s1\"}\n");
assert_test('claude session', $r->sessionId === 's1');

$r = JsonOutput::parseClaudeCodeJson("{\"type\":\"assistant\",\"message\":{\"content\":[{\"type\":\"text\",\"text\":\"hi\"}]}}\n");
assert_test('claude assistant', $r->events[0]->eventType === 'assistant' && $r->finalText === 'hi');

$r = JsonOutput::parseClaudeCodeJson("{\"type\":\"stream_event\",\"event\":{\"type\":\"content_block_delta\",\"delta\":{\"type\":\"text_delta\",\"text\":\"yo\"}}}\n");
assert_test('claude text delta', $r->events[0]->eventType === 'text_delta' && $r->events[0]->text === 'yo');

$r = JsonOutput::parseClaudeCodeJson("{\"type\":\"tool_use\",\"tool_name\":\"read\",\"tool_input\":{\"a\":1}}\n");
assert_test('claude tool use', $r->events[0]->eventType === 'tool_use' && $r->events[0]->toolCall->name === 'read');

$r = JsonOutput::parseClaudeCodeJson("{\"type\":\"tool_result\",\"tool_use_id\":\"t1\",\"content\":\"out\",\"is_error\":false}\n");
assert_test('claude tool result', $r->events[0]->toolResult->toolCallId === 't1' && $r->events[0]->toolResult->content === 'out');

$r = JsonOutput::parseClaudeCodeJson("{\"type\":\"result\",\"subtype\":\"success\",\"result\":\"done\",\"cost_usd\":0.1,\"duration_ms\":500}\n");
assert_test('claude result success', $r->finalText === 'done' && $r->costUsd === 0.1 && $r->durationMs === 500);

$r = JsonOutput::parseKimiJson("{\"role\":\"assistant\",\"content\":\"hello\"}\n");
assert_test('kimi assistant', $r->events[0]->eventType === 'assistant' && $r->finalText === 'hello');

$r = JsonOutput::parseKimiJson("{\"role\":\"assistant\",\"content\":\"\",\"tool_calls\":[{\"id\":\"1\",\"function\":{\"name\":\"bash\",\"arguments\":\"{}\"}}]}\n");
$found = false;
foreach ($r->events as $ev) { if ($ev->eventType === 'tool_call' && $ev->toolCall->name === 'bash') $found = true; }
assert_test('kimi tool call', $found);

$r = JsonOutput::parseJsonOutput("{\"response\":\"ok\"}\n", "opencode");
assert_test('dispatch opencode', $r->schemaName === 'opencode' && $r->finalText === 'ok');

$r = JsonOutput::parseJsonOutput("", "unknown");
assert_test('unknown schema error', strlen($r->error) > 0);

$mock = new ParsedJsonOutput();
$mock->schemaName = 'test';
$mock->events = [
    (function() { $e = new \Call\Coding\Clis\JsonEvent(); $e->eventType = 'text'; $e->text = 'hello'; return $e; })(),
    (function() { $e = new \Call\Coding\Clis\JsonEvent(); $e->eventType = 'assistant'; $e->text = 'world'; return $e; })(),
];
assert_test('render text events', JsonOutput::renderParsed($mock) === "hello\nworld");

$mock2 = new ParsedJsonOutput();
$mock2->schemaName = 'test';
$mock2->finalText = 'fallback';
assert_test('render fallback', JsonOutput::renderParsed($mock2) === 'fallback');

echo "\nResults: $passed passed, $failed failed\n";
exit($failed > 0 ? 1 : 0);
